aeneas/src/arm64/SsaArm64Gen.v3

```diff
@@ -129,87 +224,288 @@ class SsaArm64Gen extends SsaMachGen {
 
 		def am = getAM(opcode);
 		match (getAM(opcode)) {
-			AM_R_R_I_I => {
-				def imm = i12.view(toB32(toImm(a[2])));
-				def lsl12 = u1.view(getArg1(opcode));
-				assemble_r_r_i12_u1(toGpr(a[0]), toGpr(a[1]), imm, lsl12, opcode);
+			AM_R_R => assemble_r_r(toGpr(a[0]), toGpr(a[1]), opcode);
+			AM_R_R_R => assemble_r_r_r(toGpr(a[0]), toGpr(a[1]), toGpr(a[2]), opcode);
+			AM_R_R_I12_U1 => {
+				assemble_r_r_i12_u1(toGpr(a[0]), toGpr(a[1]), toFixedInt<i12>(a[2]),
+					u1.!(getArg1(opcode)), opcode);
+			}
+			AM_R_R_R_SH_U5 => {
+				assemble_r_r_r_sh_u5(toGpr(a[0]), toGpr(a[1]), toGpr(a[2]),
+					toRegShift(getArg1(opcode)), u5.!(getArg2(opcode)), opcode);
 			}
-			AM_R_R_R_SH_I => {
-				def sh = toRegShift(getArg1(opcode));
-				def imm = u6.view(getArg2(opcode));
-				assemble_r_r_r_sh_u6(toGpr(a[0]), toGpr(a[1]), toGpr(a[2]), sh, imm, opcode);
+			AM_R_R_R_SH_U6 => {
+				assemble_r_r_r_sh_u6(toGpr(a[0]), toGpr(a[1]), toGpr(a[2]),
+					toRegShift(getArg1(opcode)), u6.!(getArg2(opcode)), opcode);
 			}
+			AM_R_R_R_EX_U3 => {
+				assemble_r_r_r_ex_u3(toGpr(a[0]), toGpr(a[1]), toGpr(a[2]),
+					toDataRegExtend(getArg1(opcode)), u3.!(getArg2(opcode)), opcode);
+			}
+			AM_R_I16 => assemble_r_i16(toGpr(a[0]), toFixedInt<i16>(a[1]), opcode);
+			AM_R_I16_U1 => assemble_r_i16_u1(toGpr(a[0]), toFixedInt<i16>(a[1]),
+				u1.!(getArg1(opcode)), opcode);
+			AM_R_I16_U2 => assemble_r_i16_u2(toGpr(a[0]), toFixedInt<i16>(a[1]),
+				u2.!(getArg1(opcode)), opcode);
+			AM_M_R => assemble_r_r_i32(toGpr(a[1]), Regs.SP, toRspOffset(a[0]), opcode);
+			AM_R_M => assemble_r_r_i32(toGpr(a[0]), Regs.SP, toRspOffset(a[1]), opcode);	
+			AM_NONE => assemble_none(opcode, a);
 			_ => return context.fail1("unknown addressing mode %d", am);
 		}
 	}
+	def getOutput() -> ArchInstrBuffer {
+		if (out != null) return out;
+		return out = Arm64InstrBuffer.new(this, context.prog, Arm64RegSet.SET);
+	}
+
+	// ========================================
+	// Helper Functions
+	// ========================================
 
+	// ========================================
+	// Assembling Helpers
+	// ========================================	
+
+	def assemble_none(opcode: int, a: Array<Operand>) {
+		match (getCode(opcode)) {
+			I_BL => {
+				var target: bool, outgoing: MachCallConv, livepoint = -1;
+				for (o in a) {
+					match (o) {
```

Comment:
It seems like `BL` doesn't really fit under the "none" addressing mode. If you're switching over operands, then it's better to encode that into the addressing mode.